anonymous_helpline_chatbot/core/invitations.py

```diff
@@ -1,89 +1,131 @@
-from sys import stderr
-from traceback import format_exc
-
-import telebot
 from typing import Callable, Any
 
+from .db_connector import DatabaseConnectionPool, cursor_type
 from .users import UsersController
 from .conversations import ConversationsController
 
 
 class InvitationsController:
-    def __init__(self, users_controller: UsersController, conversations_controller: ConversationsController,
+    def __init__(self, database_connection_pool: DatabaseConnectionPool,
+                 users_controller: UsersController, conversations_controller: ConversationsController,
                  send_invitation_callback: Callable[[int, int, str], int],
                  delete_invitation_callback: Callable[[int, int], Any]):
+        self._conn_pool = database_connection_pool
+
         self.users_controller = users_controller
         self.conversations_controller = conversations_controller
 
         # Whenever a client requests a conversation, all the <b>free</b> operators get a message which invites them to
         # start chatting with that client. Whenever an operator accepts the invitation, all the messages which invite to
         # the conversation with that client are deleted, and the conversation begins between the client and the operator
-        # who accepted the invitation
-
-        # `_operators_invitations_messages` is used for storing sent invitations messages for being able to delete them
-        # later. It is a dictionary from telegram client id to a list of tuples of telegram operator chat id and
-        # telegram id of a message, which invites the operator to join a conversation with the client
-        # (simpler `{client_id: [(operator_id, message_id)]}`)
-        self._operators_invitations_messages = {}
+        # who accepted the invitation.
+        # However, the above behavior will be altered soon
 
         self.send_invitation_callback = send_invitation_callback
         self.delete_invitation_callback = delete_invitation_callback
 
-    def invite_operators(self, tg_client_id: int) -> int:
+    def _invite_operator_to_client(self, cursor: cursor_type, operator_chat_id: int, client_chat_id: int) -> None:
+        client_local_id = self.users_controller.get_local_id(client_chat_id)
+
+        sent_message_id = self.send_invitation_callback(operator_chat_id, client_chat_id,
+                                                        f"Пользователь №{client_local_id} хочет побеседовать. Нажмите "
+                                                        "кнопку ниже, чтобы стать его оператором")
+
+        # Unless message sending failed for some internal front-end reason, store invitation in the database
+        if sent_message_id is not None:
+            cursor.execute("INSERT INTO sent_invitations(operator_chat_id, client_chat_id, invitation_message_id) "
+                           "VALUES (%s, %s, %s)",
+                           (operator_chat_id, client_chat_id, sent_message_id))
```

Comment:
In case this method is called from two parallel threads with the same operator/client ids, it is possible that both messages get sent by the front-end, then one thread stores the message it has sent in the database, but the other thread is unable to do this because the scheme forbids multiple invitations for a pair (operator_chat_id, client_chat_id).

In this situation an "invitation leak", as I call it, happens: the operator gets two invitations to the same client, but only one of those invitations is tracked in the database and can be deleted. This means that the other invitation will stay in the operator's chat with the bot forever. This is a very serious issue, because, among other things, it allows the operator to start a conversation with the client whenever he wants it (even if the client doesn't request conversation at all).

That's why I want the leaked invitation (i.e. invitation which failed to be inserted to the database) to be deleted immediately with `self.delete_invitation_callback`